clustered/integration-test/src/test/java/org/ehcache/clustered/util/ParallelTestCluster.java

```diff
@@ -0,0 +1,201 @@
+
+  public ParallelTestCluster(Cluster cluster) {
+    this.cluster = cluster;
+
+    IClusterControl underlyingControl = cluster.getClusterControl();
+    this.control = new IClusterControl() {
+      @Override
+      public void waitForActive() throws Exception {
+        underlyingControl.waitForActive();
+      }
+
+      @Override
+      public void waitForRunningPassivesInStandby() throws Exception {
+        underlyingControl.waitForRunningPassivesInStandby();
+      }
+
+      @Override
+      public void startOneServer() {
+        request(ClusterTask.START_ONE_SERVER);
+      }
+
+      @Override
+      public void startOneServerWithConsistency() {
+        request(ClusterTask.START_ONE_SERVER_WITH_CONSISTENCY);
+      }
+
+      @Override
+      public void startAllServers() {
+        request(ClusterTask.START_ALL_SERVERS);
+      }
+
+      @Override
+      public void startAllServersWithConsistency() {
+        request(ClusterTask.START_ALL_SERVERS_WITH_CONSISTENCY);
+      }
+
+      @Override
+      public void terminateActive() {
+        request(ClusterTask.TERMINATE_ACTIVE);
+      }
+
+      @Override
+      public void terminateOnePassive() {
+        request(ClusterTask.TERMINATE_ONE_PASSIVE);
+      }
+
+      @Override
+      public void terminateAllServers() {
+        request(ClusterTask.TERMINATE_ALL_SERVERS);
+      }
+
+      private void request(ClusterTask task) {
+        try {
+          if (nextTask.compareAndSet(null, task)) {
+            activeCycle.awaitAdvanceInterruptibly(activeCycle.arrive());
+            nextTask.getAndSet(null).accept(underlyingControl);
+            activeCycle.awaitAdvanceInterruptibly(activeCycle.arrive());
+          } else {
+            ClusterTask requestedTask = nextTask.get();
+            if (requestedTask.equals(task)) {
+              activeCycle.awaitAdvanceInterruptibly(activeCycle.arrive());
+              activeCycle.awaitAdvanceInterruptibly(activeCycle.arrive());
+            } else {
+              throw new AssertionError("Existing requested task is " + requestedTask);
+            }
+          }
+        } catch (InterruptedException e) {
+          throw new AssertionError(e);
+        }
+      }
+    };
+  }
+
+  @Override
+  public URI getConnectionURI() {
+    return cluster.getConnectionURI();
+  }
+
+  @Override
+  public String[] getClusterHostPorts() {
+    return cluster.getClusterHostPorts();
+  }
+
+  @Override
+  public Connection newConnection() throws ConnectionException {
+    return cluster.newConnection();
+  }
+
+  @Override
+  public IClusterControl getClusterControl() {
+    return control;
+  }
+
+  @Override
+  public Statement apply(Statement base, Description description) {
+    return cluster.apply(base, description);
+  }
+
+  public TestRule watchdog() {
+    return new TestWatcher() {
+
+      @Override
+      protected void starting(Description description) {
+        membership.register();
+        try {
+          Thread.sleep(100);
```

Comment:
why we need a sleep ?

Reply:
We don't *need* a sleep... we sleep to give some time for more participants to show up for the test cycle.